Added support for multiple test frameworks

erik-inkapool opened the pull request:
With this we can support MSTest and NUnit by just looking at the `testRunner` entry in project.json since the moniker there will be identical to the one used in app.config.

I haven't tested this extensively and MSTest produces fairly ugly test names (see code comment) but the sample builds and tests fine on both those test frameworks.

Also editorconfig seems to be doing its job. Yay pretty diffs!

erik-inkapool committed:
Added support for multiple test frameworks

erik-inkapool commented on src/SpecFlow.NetCore/Fixer.cs:
```diff
@@ -56,22 +57,35 @@ private void DeleteFakeCsProj(FileInfo fakeCsproj)
 
+		private static string FixMsTest(string content)
+		{
+			content = Regex.Replace(content, ".*Microsoft.VisualStudio.TestTools.UnitTesting.Description.*", "");
```
There doesn't seem to be any support for this or a similar attribute in latest for netcore

stajs replied:
Is that a problem?

erik-inkapool replied:
I think it's the cause of the ugly mstest names, but i dont know how to fix it or if a fix exist as I found no similar attribute.

erik-inkapool commented on src/SpecFlow.NetCore/AppConfig.cs:
```diff
@@ -52,16 +54,34 @@ public static AppConfig CreateIn(DirectoryInfo directory)
+			if (foundProjectJson != null)
+			{
+				dynamic projectJson = JsonConvert.DeserializeObject(File.ReadAllText(foundProjectJson.FullName));
+				if (projectJson.testRunner != null)
```
Just realised dynamics don't work like this and this will likely throw. Will fix tomorrow

stajs replied:
OK, will wait til you are happy before merging.

stajs commented on src/SpecFlow.NetCore/AppConfig.cs:
```diff
@@ -52,16 +54,34 @@ public static AppConfig CreateIn(DirectoryInfo directory)
 		}
 
-		public void Validate()
+		private static string GuessUnitTestProvider(DirectoryInfo directory)
```
I'm in two minds about guessing the provider. I like the convenience but I also like being explicit; if we can take it as an argument then we don't need to depend on `Newtonsoft.Json`.

erik-inkapool replied:
Yeah you have a good point, but you can always force provider by changing the generated app config. Currently we take in a path argument, do we still need to support that?

stajs replied:
Yes, a good point that you can force the provider after the `app.config` is generated. I'm happy enough with guessing for now. Can always change it later if anyone complains about unneeded dependencies - I suspect most people will be using `Newtonsoft.Json` already!

The path argument is vestigial and can be removed, or replaced with a provider argument in the future if required.

stajs commented on samples/VS2015/SpecFlow 2.1.0/net461/Sample.Website.Tests/project.json:
```diff
@@ -15,7 +15,7 @@
 
 	"tools": {
-		"SpecFlow.NetCore": "1.0.0-rc4"
+		"SpecFlow.NetCore": "1.0.2"
```
I'll keep it RC until it's stable.

erik-inkapool replied:
that's just for my testing, sorry. Nuget keeps caching the tool so I have to update versions all the time. Do you know of a workaround?

stajs commented on src/SpecFlow.NetCore/Fixer.cs:
```diff
@@ -56,22 +57,35 @@ private void DeleteFakeCsProj(FileInfo fakeCsproj)
-				content = content.Replace(" : Xunit.IUseFixture<", " : Xunit.IClassFixture<");
-				content = content.Replace("[Xunit.Extensions", "[Xunit");
+				content = FixXUnit(content);
+				content = FixMsTest(content);
```
Micro-optimisation: could hold on to the provider and only fix tests for that provider.

erik-inkapool replied:
Yeah the mstest one is a regex so might be appropriate. I'll let you decide on argument/guess first and I'll fix this once we have a decision there.

stajs commented:
Is this PR waiting on anything else or can it be merged?

stajs commented:
Unfortunately there are some conflicts now. Is this worth trying to merge?

erik-inkapool commented:
Hey Stajs, I'll try to fix this during the weekend. I'd like to put in a commandline argument as well before we merge it if you don't mind.

stajs commented:
Sounds good, should be relatively easy to support arguments now!

erik-inkapool committed:
Merge branch 'master' into 30

erik-inkapool committed:
Added argument for test runner

Also added a feature to always replace test runner part of app.config if
app.config exists.

Also fixed a small bug if there are spaces in the project path

erik-inkapool commented on src/SpecFlow.NetCore/AppConfig.cs:
```diff
@@ -18,59 +21,117 @@ internal class AppConfig
 		}
 
-		public void Validate()
+		public void ChangeTestProviderIfNeeded(string testRunner)
```
This is a pretty wordy and brittle method (if it doesn't find the exact app config stuff it does nothing silently). It also means it is no longer possible to change the app.config testRunner as we will overwrite it. Probably not a big deal as test runner is now specified using the arguments instead.

stajs replied:
Hmm, I think I prefer considering this as part of the validation rather than modifying the `app.config` which will probably go unnoticed. Will let it settle to see how it feels...

erik-inkapool commented on src/SpecFlow.NetCore/Fixer.cs:
```diff
@@ -85,26 +89,45 @@ private void DeleteFakeCsProj(FileInfo fakeCsproj)
 		{
 			// Credit: http://www.marcusoft.net/2010/12/specflowexe-and-mstest.html
-			var arguments = $"generateall {csproj} /force /verbose";
+			var arguments = $@"generateall ""{csproj}"" /force /verbose";
```
Fixes a bug where if you try to generate with a space in path it breaks.

stajs replied:
Nice catch, thanks!

erik-inkapool commented on src/SpecFlow.NetCore/project.json:
```diff
@@ -1,5 +1,5 @@
 ﻿{
-	"version": "1.0.0-rc5",
+	"version": "1.0.0-rc4", // TODO: update version
```
Leaving this to you, @stajs

stajs replied:
👍

erik-inkapool commented:
If you feel it's ready, you can merge this now :)

stajs commented on src/SpecFlow.NetCore/AppConfig.cs:
```diff
@@ -18,59 +21,118 @@ internal class AppConfig
 		}
 
-		public static AppConfig CreateIn(DirectoryInfo directory)
+		public static AppConfig CreateIn(DirectoryInfo directory, out string usedTestRunner, string testRunner = null)
```
I think we can get away without an `out` parameter, and just use the `TestFramework` property.

erik-inkapool replied:
I like that too, nice solution

stajs commented on src/SpecFlow.NetCore/Args.cs:
```diff
@@ -56,6 +62,7 @@ public Args(string[] args)
 
 			WriteLine("SpecFlowPath: " + SpecFlowPath);
 			WriteLine("WorkingDirectory: " + WorkingDirectory.FullName);
+			WriteLine("Test runner: " + WorkingDirectory.FullName);
```
I'll fix this.

stajs commented:
@erik-lundgren Thanks for work on this!
